[DSE] Update dereferenceable attributes when adjusting memintrinsic ptr

bjope opened the pull request:
Consider IR like this
  call void @llvm.memset.p0.i64(ptr dereferenceable(28) %p, i8 0, i64 28, i1 false)
  store i32 1, ptr %p

In the past it has been optimized like this:
  %p2 = getelementptr inbounds i8, ptr %p, i64 4
  call void @llvm.memset.p0.i64(ptr dereferenceable(28) %p2, i8 0, i64 24, i1 false)
  store i32 1, ptr %p

As the input IR doesn't guarantee that it is OK to deref 28 bytes starting at the adjusted pointer %p2 the transformation has been a bit flawed.

With this patch we make sure to drop any dereferenceable/dereferenceable_or_null attributes when doing such transforms. An alternative would have been to adjust the amount of dereferenceable bytes, but since a memset with a constant length already implies dereferenceability by itself it is simpler to just drop the attributes.

The new filtering of attributes is done using a helper that only keep attributes that we explicitly handle. For the adjusted mem instrinsic pointers that currently involve "NonNull", "NoUndef" and "Alignment" (when the alignment is known to be fulfilled also after offsetting the pointer).

Fixes #115976

bjope committed:
[DSE] Update dereferenceable attributes when adjusting memintrinsic ptr

Consider IR like this
  call void @llvm.memset.p0.i64(ptr dereferenceable(28) %p, i8 0, i64 28, i1 false)
  store i32 1, ptr %p

It has been optimized like this:
  %p2 = getelementptr inbounds i8, ptr %p, i64 4
  call void @llvm.memset.p0.i64(ptr dereferenceable(28) %p2, i8 0, i64 24, i1 false)
  store i32 1, ptr %p

As the input IR doesn't guarantee that it is OK to deref 28 bytes
starting at the adjusted pointer %p2 the transformation has been
a bit flawed.

With this patch we make sure to also adjust the size of any
dereferenceable/dereferenceable_or_null attributes when doing the
transform in tryToShorten (when adjusting the start pointer). So now
we will get dereferenceable(24) in the example above.

bjope committed:
fixup formatting

nikic commented on llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:
```diff
@@ -563,6 +563,22 @@ static void shortenAssignment(Instruction *Inst, Value *OriginalDest,
   for_each(LinkedDVRAssigns, InsertAssignForOverlap);
 }
 
+// Helper to trim or drop any dereferencable/dereferencable_or_null attributes
```
```suggestion
// Helper to trim or drop any dereferenceable/dereferenceable_or_null attributes
```

nikic commented on llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:
```diff
@@ -563,6 +563,22 @@ static void shortenAssignment(Instruction *Inst, Value *OriginalDest,
 }
 
+// Helper to trim or drop any dereferencable/dereferencable_or_null attributes
+// for a given argument, based on the new access being restricted to derefence
```
```suggestion
// for a given argument, based on the new access being restricted to dereference
```

nikic commented on llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:
```diff
@@ -563,6 +563,22 @@ static void shortenAssignment(Instruction *Inst, Value *OriginalDest,
+// Helper to trim or drop any dereferencable/dereferencable_or_null attributes
+// for a given argument, based on the new access being restricted to derefence
+// bytes in the range [Offset, Offset+Size).
+static void trimDereferencableAttrs(AnyMemIntrinsic *Intrinsic, unsigned Arg,
```
```suggestion
static void trimDereferenceableAttrs(AnyMemIntrinsic *Intrinsic, unsigned Arg,
```

nikic commented on llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:
```diff
@@ -563,6 +563,22 @@ static void shortenAssignment(Instruction *Inst, Value *OriginalDest,
+                                    uint64_t Offset, uint64_t Size) {
+  uint64_t End = Offset + Size;
+  if (Intrinsic->getParamDereferenceableBytes(Arg) >= End)
+    Intrinsic->addDereferenceableParamAttr(Arg, Size);
```
If you're going to actually adjust the attributes, I'd expect to see something like "OldDerefenceableBytes - Offset" here, so that information about larger dereferenceability than the memset length is preserved.

But really, I'd consider to just completely drop the attributes in this transform. memset with a constant length already implies dereferenceability by itself.

bjope replied:
Ok, thanks. I wasn't quite sure if dropping the attributes could have some downsides, but just dropping them is simpler so I like that solution. So I've updated the patch to just drop the attributes instead of trying to adjust them.

efriedma-quic replied:
Probably we should drop all attributes here, so we don't need to adjust this whenever we add a new attribute to LLVM.

nikic replied:
Dropping all attributes would also lose alignment, which is correctly adjusted and might not be recoverable...

bjope replied:
I see some alternatives:
1) Instead of having an explicit set of attributes to drop we could use prepare a set of attributes that is safe to keep. And then use that for the filtering. That would more future proof in case new attributes are added.
2) Another idea is to add some kind of support in Attributes.h (or similar). Such as a more centralized helper method to "filter attributes based on access being narrowed" (and then it might be easier to find and update that when adding new attributes). This could also just be an AttributeMask similar to getUBImplyingAttributes().

bjope committed:
Simplify the solution by just dropping the dereferenceable attrs
when there is a risk that they would become incorrect if neither
adjusting nor dropping them.

dtcxzyw commented:
Does it fix https://github.com/llvm/llvm-project/issues/115976?

nikic reviewed (commented):
This introduces one regression: https://llvm.godbolt.org/z/1EjrE9xM6

It's due to this check: https://github.com/llvm/llvm-project/blob/2bffa5bf7abd4fc7b84f14d029a97c49f4b61130/llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp#L2213

We should be doing the check in IntersectAttrs mode.

nikic commented:
Regression should be fixed by https://github.com/llvm/llvm-project/pull/125190.

bjope commented:
> Does it fix #115976?

Yes it should. I've added as "Fixes" in the description.

bjope committed:
Change the logic to drop all attributes that we do not explicitly
handle. That should be more future proof as we automatically would
drop any new attributes being added.

bjope commented:
Ping

dtcxzyw reviewed (approved):
LGTM. Thank you!
